### ISSUE TITLE
关于sage attention和flash attention的速度对比

### ISSUE THREAD

**Issue by cslvjt** [thu-ml/SageAttention#343]
感谢作者为社区做的贡献，非常有用！我最近在对比sageAttention和flashattention的速度时，发现当输入张量中序列长度只有超过2048时，sageattention的推理速度比flash attention更快。作者能否解释一下原因呢？非常感谢！

<img width="1372" height="528" alt="Image" src="https://github.com/user-attachments/assets/82521667-bf9d-49fb-abec-738643b3c914" />
环境：L40 ｜ CUDA 12.8 ｜sageAttention 2.2
代码：

"""
FlashAttention vs SageAttention 速度测试
测试不同 seq_length 下的性能对比
"""

import torch
import torch.nn.functional as F
import time
try:
    from sageattention import sageattn
    SAGE_AVAILABLE = True
except ImportError:
    SAGE_AVAILABLE = False
    print("Warning: sageattention not available")
def benchmark_attention(batch_size=32, num_heads=5, head_dim=64, warmup=3, iters=10):
    """
    测试不同 seq_length 下 FlashAttention 和 SageAttention 的速度
    
    输入形状: [batch_size, seq_length, num_heads, head_dim]
    """
    print("=" * 90)
    print(f"Attention Speed Benchmark")
    print(f"Shape: [{batch_size}, seq_length, {num_heads}, {head_dim}]")
    print(f"Warmup: {warmup}, Iterations: {iters}")
    print("=" * 90)
    print(f"{'seq_len':>10} | {'FlashAttn (ms)':>15} | {'SageAttn (ms)':>15} | {'Speedup':>10}")
    print("-" * 90)
    
    results = []
    
    # seq_length 从 16 到 2048，每次乘以 2
    seq_length = 16
    while seq_length <= 2048:
        # 创建输入数据
        q = torch.randn(batch_size, seq_length, num_heads, head_dim, 
                       device="cuda", dtype=torch.float16).contiguous()
        k = torch.randn(batch_size, seq_length, num_heads, head_dim, 
                       device="cuda", dtype=torch.float16).contiguous()
        v = torch.randn(batch_size, seq_length, num_heads, head_dim, 
                       device="cuda", dtype=torch.float16).contiguous()
        
        # ============ FlashAttention (scaled_dot_product_attention) ============
        # 需要转换为 (batch, num_heads, seq_len, head_dim) 格式
        q_flash = q.permute(0, 2, 1, 3).contiguous()
        k_flash = k.permute(0, 2, 1, 3).contiguous()
        v_flash = v.permute(0, 2, 1, 3).contiguous()
        
        # Warmup
        for _ in range(warmup):
            out = F.scaled_dot_product_attention(q_flash, k_flash, v_flash, is_causal=False)
            torch.cuda.synchronize()
        
        # Benchmark
        torch.cuda.synchronize()
        start = time.time()
        for _ in range(iters):
            out = F.scaled_dot_product_attention(q_flash, k_flash, v_flash, is_causal=False)
            torch.cuda.synchronize()
        flash_time = (time.time() - start) / iters * 1000  # ms
        
        # ============ SageAttention ============
        if SAGE_AVAILABLE:
            # Warmup
            for _ in range(warmup):
                out = sageattn(q, k, v, tensor_layout="NHD", is_causal=False)
                torch.cuda.synchronize()
            
            # Benchmark
            torch.cuda.synchronize()
            start = time.time()
            for _ in range(iters):
                out = sageattn(q, k, v, tensor_layout="NHD", is_causal=False)
                torch.cuda.synchronize()
            sage_time = (time.time() - start) / iters * 1000  # ms
            
            speedup = flash_time / sage_time
            print(f"{seq_length:>10} | {flash_time:>15.3f} | {sage_time:>15.3f} | {speedup:>9.2f}x")
            results.append((seq_length, flash_time, sage_time, speedup))
        else:
            print(f"{seq_length:>10} | {flash_time:>15.3f} | {'N/A':>15} | {'N/A':>10}")
            results.append((seq_length, flash_time, None, None))
        
        # 清理显存
        del q, k, v, q_flash, k_flash, v_flash
        torch.cuda.empty_cache()
        
        # seq_length 乘以 2
        seq_length *= 2
    
    # 打印汇总
    print("=" * 90)
    print("Summary:")
    if SAGE_AVAILABLE and results:
        avg_speedup = sum(r[3] for r in results if r[3]) / len([r for r in results if r[3]])
        print(f"Average Speedup: SageAttention is {avg_speedup:.2f}x faster than FlashAttention")
    print("=" * 90)
    
    return results

if __name__ == "__main__":
    benchmark_attention(
        batch_size=32,
        num_heads=5,
        head_dim=64,
        warmup=3,
        iters=10
    )
